File: odyssey-api/src/utils/validation.ts
import { z } from "zod";
import { AppError } from "../middleware/error-handler";

export const syncEntrySchema = z.object({
  entryDate: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),

  // Encrypted fields (base64 ciphertext, optional)
  journalEntry: z.string().nullable().optional(),
  gratitude: z.string().nullable().optional(),
  win: z.string().nullable().optional(),
  tension: z.string().nullable().optional(),
  singleWordFeeling: z.string().nullable().optional(),
  latitude: z.string().nullable().optional(),
  longitude: z.string().nullable().optional(),
  city: z.string().nullable().optional(),
  state: z.string().nullable().optional(),
  country: z.string().nullable().optional(),

  // Plaintext fields
  feeling: z.number().int().min(1).max(10),
  sleepQuality: z.number().int().min(1).max(10),
  feelingColorHex: z.string(),
  drinks: z.number().int().min(0),
  stepCount: z.number().int().nullable().optional(),
  walkingDistanceMeters: z.number().nullable().optional(),
  sleepHours: z.number().nullable().optional(),
  screenTimeSeconds: z.number().nullable().optional(),
  pickups: z.number().int().nullable().optional(),

  createdAt: z.string(),
  updatedAt: z.string(),
});

export const syncUploadSchema = z.object({
  entries: z.array(syncEntrySchema).min(1).max(100),
});

export const keyBackupSchema = z.object({
  encryptedKeyData: z.string().min(1),
  keyDerivationSalt: z.string().nullable().optional(),
});

export type SyncEntry = z.infer<typeof syncEntrySchema>;
export type SyncUploadPayload = z.infer<typeof syncUploadSchema>;
export type KeyBackupPayload = z.infer<typeof keyBackupSchema>;

// --- Post-parse validators ---

const encoder = new TextEncoder();
const HEX_COLOR_RE = /^#[0-9A-Fa-f]{6}$/;
const MAX_ENCRYPTED_FIELD_BYTES = 50_000;
const MAX_ENCRYPTED_KEY_DATA_BYTES = 10_240;
const MAX_KEY_DERIVATION_SALT_BYTES = 500;
const TIMESTAMP_DRIFT_MS = 48 * 60 * 60 * 1000; // 48 hours

export function validateByteLength(
  value: string | null | undefined,
  name: string,
  maxBytes: number
): void {
  if (value == null) return;
  const len = encoder.encode(value).byteLength;
  if (len > maxBytes) {
    throw AppError.validation(
      `${name} exceeds maximum length of ${maxBytes} bytes (got ${len})`
    );
  }
}

export function validateTimestamp(ts: string, fieldName: string): void {
  const parsed = Date.parse(ts);
  if (isNaN(parsed)) {
    throw AppError.validation(`Invalid ${fieldName} timestamp: ${ts}`);
  }
  const diff = Math.abs(Date.now() - parsed);
  if (diff > TIMESTAMP_DRIFT_MS) {
    throw AppError.validation(
      `${fieldName} is more than 48 hours from server time`
    );
  }
}

export function validateEntryDateRange(date: string): void {
  const parsed = new Date(date + "T00:00:00Z");
  if (isNaN(parsed.getTime())) {
    throw AppError.validation(`Invalid entry_date: ${date}`);
  }

  const today = new Date();
  today.setUTCHours(0, 0, 0, 0);

  const oneDayFuture = new Date(today);
  oneDayFuture.setUTCDate(oneDayFuture.getUTCDate() + 1);

  const twoYearsPast = new Date(today);
  twoYearsPast.setUTCDate(twoYearsPast.getUTCDate() - 365 * 2);

  if (parsed > oneDayFuture) {
    throw AppError.validation(
      `entry_date ${date} is more than 1 day in the future`
    );
  }
  if (parsed < twoYearsPast) {
    throw AppError.validation(
      `entry_date ${date} is more than 2 years in the past`
    );
  }
}

export function validateHexColor(color: string): void {
  if (!HEX_COLOR_RE.test(color)) {
    throw AppError.validation(
      `Invalid hex color format: ${color}. Expected #RRGGBB`
    );
  }
}

export function validateSyncEntry(entry: SyncEntry): void {
  validateTimestamp(entry.createdAt, "createdAt");
  validateTimestamp(entry.updatedAt, "updatedAt");
  validateEntryDateRange(entry.entryDate);
  validateHexColor(entry.feelingColorHex);

  // Validate encrypted field byte lengths
  const encryptedFields: [string | null | undefined, string][] = [
    [entry.journalEntry, "journalEntry"],
    [entry.gratitude, "gratitude"],
    [entry.win, "win"],
    [entry.tension, "tension"],
    [entry.singleWordFeeling, "singleWordFeeling"],
    [entry.latitude, "latitude"],
    [entry.longitude, "longitude"],
    [entry.city, "city"],
    [entry.state, "state"],
    [entry.country, "country"],
  ];
  for (const [value, name] of encryptedFields) {
    validateByteLength(value, name, MAX_ENCRYPTED_FIELD_BYTES);
  }
}

export function validateKeyBackup(data: KeyBackupPayload): void {
  validateByteLength(
    data.encryptedKeyData,
    "encryptedKeyData",
    MAX_ENCRYPTED_KEY_DATA_BYTES
  );
  validateByteLength(
    data.keyDerivationSalt,
    "keyDerivationSalt",
    MAX_KEY_DERIVATION_SALT_BYTES
  );
}

File: deprecated/odyssey-server/src/validation.rs
use chrono::{DateTime, NaiveDate, Utc};

use crate::error::AppError;

/// Validate entry_date matches YYYY-MM-DD format.
pub fn validate_entry_date(date: &str) -> Result<(), AppError> {
    let re = regex_lite::Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();
    if !re.is_match(date) {
        return Err(AppError::Validation(format!(
            "Invalid entry_date format: {date}. Expected YYYY-MM-DD"
        )));
    }
    Ok(())
}

/// Parse and validate an ISO 8601 timestamp, rejecting values >48h from server time.
pub fn validate_timestamp(ts: &str, field_name: &str) -> Result<DateTime<Utc>, AppError> {
    let parsed: DateTime<Utc> = ts.parse().map_err(|e| {
        AppError::Validation(format!("Invalid {field_name} timestamp: {e}"))
    })?;

    let now = Utc::now();
    let diff = (now - parsed).abs();
    let max_drift = chrono::Duration::hours(48);

    if diff > max_drift {
        return Err(AppError::Validation(format!(
            "{field_name} is more than 48 hours from server time"
        )));
    }

    Ok(parsed)
}

/// Parse and validate an entry date string, rejecting dates >1 day in the future
/// or >2 years in the past.
pub fn validate_entry_date_range(date: &str) -> Result<NaiveDate, AppError> {
    let parsed = date.parse::<NaiveDate>().map_err(|e| {
        AppError::Validation(format!("Invalid entry_date: {e}"))
    })?;

    let today = Utc::now().date_naive();
    let one_day_future = today + chrono::Duration::days(1);
    let two_years_past = today - chrono::Duration::days(365 * 2);

    if parsed > one_day_future {
        return Err(AppError::Validation(format!(
            "entry_date {date} is more than 1 day in the future"
        )));
    }

    if parsed < two_years_past {
        return Err(AppError::Validation(format!(
            "entry_date {date} is more than 2 years in the past"
        )));
    }

    Ok(parsed)
}

/// Validate feeling is 1-10.
pub fn validate_feeling(val: i32) -> Result<(), AppError> {
    if !(1..=10).contains(&val) {
        return Err(AppError::Validation(format!(
            "feeling must be between 1 and 10, got {val}"
        )));
    }
    Ok(())
}

/// Validate sleep_quality is 1-10.
pub fn validate_sleep_quality(val: i32) -> Result<(), AppError> {
    if !(1..=10).contains(&val) {
        return Err(AppError::Validation(format!(
            "sleepQuality must be between 1 and 10, got {val}"
        )));
    }
    Ok(())
}

/// Validate drinks >= 0.
pub fn validate_drinks(val: i32) -> Result<(), AppError> {
    if val < 0 {
        return Err(AppError::Validation(format!(
            "drinks must be >= 0, got {val}"
        )));
    }
    Ok(())
}

/// Validate that an optional encrypted string field does not exceed `max_len` bytes.
pub fn validate_encrypted_field_length(
    value: &Option<String>,
    name: &str,
    max_len: usize,
) -> Result<(), AppError> {
    if let Some(v) = value {
        if v.len() > max_len {
            return Err(AppError::Validation(format!(
                "{name} exceeds maximum length of {max_len} bytes (got {})",
                v.len()
            )));
        }
    }
    Ok(())
}

/// Validate that a required string field does not exceed `max_len` bytes.
pub fn validate_required_field_length(
    value: &str,
    name: &str,
    max_len: usize,
) -> Result<(), AppError> {
    if value.len() > max_len {
        return Err(AppError::Validation(format!(
            "{name} exceeds maximum length of {max_len} bytes (got {})",
            value.len()
        )));
    }
    Ok(())
}

/// Validate a hex color string matches `#RRGGBB` format.
pub fn validate_hex_color(color: &str) -> Result<(), AppError> {
    let re = regex_lite::Regex::new(r"^#[0-9A-Fa-f]{6}$").unwrap();
    if !re.is_match(color) {
        return Err(AppError::Validation(format!(
            "Invalid hex color format: {color}. Expected #RRGGBB"
        )));
    }
    Ok(())
}
